File: docs/examples/rust/nativeexample/examples/query.rs
use taos::*;
use chrono::Local;
use chrono::DateTime;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let dsn = "taos://localhost:6030";
    let builder = TaosBuilder::from_dsn(dsn)?;

    let taos = builder.build().await?;

    // ANCHOR: create_db_and_table
    let db = "power";
    // create database
    taos.exec_many([
        format!("CREATE DATABASE IF NOT EXISTS `{db}`"),
        format!("USE `{db}`"),
    ])
    .await?;
    println!("Create database power successfully.");

    // create super table
    taos.exec_many([
        "CREATE STABLE IF NOT EXISTS `meters` (`ts` TIMESTAMP, `current` FLOAT, `voltage` INT, `phase` FLOAT) \
            TAGS (`groupid` INT, `location` BINARY(24))",
    ]).await?;
    println!("Create stable meters successfully.");

    // ANCHOR_END: create_db_and_table

    // ANCHOR: insert_data
    let inserted = taos.exec(r#"INSERT INTO 
    power.d1001 USING power.meters TAGS(2,'California.SanFrancisco')  
    VALUES 
    (NOW + 1a, 10.30000, 219, 0.31000) 
    (NOW + 2a, 12.60000, 218, 0.33000) 
    (NOW + 3a, 12.30000, 221, 0.31000) 
    power.d1002 USING power.meters TAGS(3, 'California.SanFrancisco') 
    VALUES 
    (NOW + 1a, 10.30000, 218, 0.25000) "#).await?;

    println!("inserted: {} rows to power.meters successfully.", inserted);
    // ANCHOR_END: insert_data

    // ANCHOR: query_data
    // query data, make sure the database and table are created before
    let mut result = taos.query("SELECT ts, current, location FROM power.meters limit 100").await?;

    for field in result.fields() {
        println!("got field: {}", field.name());
    }

    let mut rows = result.rows();
    let mut nrows = 0;
    while let Some(row) = rows.try_next().await? {
        for (col, (name, value)) in row.enumerate() {
            println!(
                "[{}] got value in col {} (named `{:>8}`): {}",
                nrows, col, name, value
            );
        }
        nrows += 1;
    }
    // ANCHOR_END: query_data

    // ANCHOR: query_data_2
    // query data, make sure the database and table are created before
    #[derive(Debug, serde::Deserialize)]
    #[allow(dead_code)]
    struct Record {
        // deserialize timestamp to chrono::DateTime<Local>
        ts: DateTime<Local>,
        // float to f32
        current: Option<f32>,
        // int to i32
        voltage: Option<i32>,
        phase: Option<f32>,
        groupid: i32,
        // binary/varchar to String
        location: String,
    }

    let records: Vec<Record> = taos
        .query("select ts, current, voltage, phase, groupid, location from power.meters limit 100")
        .await?
        .deserialize()
        .try_collect()
        .await?;

    dbg!(records);
    // ANCHOR_END: query_data_2

    // ANCHOR: query_with_req_id
    let result = taos.query_with_req_id("SELECT ts, current, location FROM power.meters limit 1", 1).await?;
    for field in result.fields() {
        println!("got field: {}", field.name());
    }
    println!("query with reqId successfully");
    // ANCHOR_END: query_with_req_id
    Ok(())
}

File: source/taos-connector-dotnet/examples/WSQuery/Query.cs
using System;
using System.Text;
using TDengine.Driver;
using TDengine.Driver.Client;

namespace WSQuery
{
    internal class Query
    {
        public static void Main(string[] args)
        {
            var builder = new ConnectionStringBuilder("protocol=WebSocket;host=localhost;port=6041;useSSL=false;username=root;password=taosdata");
            using (var client = DbDriver.Open(builder))
            {
                try
                {
                    client.Exec($"create database power");
                    client.Exec($"use power");
                    client.Exec(
                        "CREATE STABLE power.meters (ts TIMESTAMP, current FLOAT, voltage INT, phase FLOAT) TAGS (groupId INT, location BINARY(24))");
                    string insertQuery =
                        "INSERT INTO " +
                        "power.d1001 USING power.meters TAGS(2,'California.SanFrancisco') " +
                        "VALUES " +
                        "('2023-10-03 14:38:05.000', 10.30000, 219, 0.31000) " +
                        "('2023-10-03 14:38:15.000', 12.60000, 218, 0.33000) " +
                        "('2023-10-03 14:38:16.800', 12.30000, 221, 0.31000) " +
                        "power.d1002 USING power.meters TAGS(3, 'California.SanFrancisco') " +
                        "VALUES " +
                        "('2023-10-03 14:38:16.650', 10.30000, 218, 0.25000) " +
                        "power.d1003 USING power.meters TAGS(2,'California.LosAngeles') " +
                        "VALUES " +
                        "('2023-10-03 14:38:05.500', 11.80000, 221, 0.28000) " +
                        "('2023-10-03 14:38:16.600', 13.40000, 223, 0.29000) " +
                        "power.d1004 USING power.meters TAGS(3,'California.LosAngeles') " +
                        "VALUES " +
                        "('2023-10-03 14:38:05.000', 10.80000, 223, 0.29000) " +
                        "('2023-10-03 14:38:06.500', 11.50000, 221, 0.35000)";
                    client.Exec(insertQuery);
                    string query = "SELECT * FROM meters";
                    using (var rows = client.Query(query))
                    {
                        while (rows.Read())
                        {
                            Console.WriteLine(
                                $"{((DateTime)rows.GetValue(0)):yyyy-MM-dd HH:mm:ss.fff}, {rows.GetValue(1)}, {rows.GetValue(2)}, {rows.GetValue(3)}, {rows.GetValue(4)}, {Encoding.UTF8.GetString((byte[])rows.GetValue(5))}");
                        }
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.ToString());
                    throw;
                }
            }
        }
    }
}